packages/design-toolkit/src/documentation/upgrades.docs.mdx

```diff
@@ -0,0 +1,374 @@
+          typeof rule.oneOf === 'object',
+      ) as RuleSetRule | undefined
+    )?.oneOf?.filter(
+      (rule) =>
+        rule != null && typeof rule === 'object' && Array.isArray(rule.use),
+    ) as RuleSetRule[] | undefined;
+
+    rules?.forEach((rule) => {
+      if (Array.isArray(rule.use)) {
+        rule.use.forEach((loader) => {
+          if (
+            loader != null &&
+            typeof loader === 'object' &&
+            loader.loader &&
+            loader.loader.includes('/css-loader/') &&
+            loader.options &&
+            typeof loader.options !== 'string' &&
+            loader.options.modules
+          ) {
+            loader.options.modules.getLocalIdent = getLocalIdent;
+          }
+        });
+      }
+    });
+
+    return config;
+  },
+};
+export default nextConfig;
+```
+
+**Vite**
+```ts
+import { generateScopedName } from './src/lib/vite';
+
+export default defineConfig({
+
+  //...your other config
+
+  css: {
+    modules: {
+      generateScopedName,
+    },
+  },
+});
+```
+
+<a id="css-modules"></a>
+### 3. Migrate to CSS modules and clsx
+Previously, we recommended using the [tailwind variants](https://www.tailwind-variants.org/) library to control css application, but because of performance and other reasons listed above, we no longer recommend that approach. While there is no formal enforcement of the these patterns, standards, consistency and code quality will allow us to move with velocity and quality.
+
+In places in your code base where you are already using [tailwind variants](https://www.tailwind-variants.org/), replace it with the following approach.
+
+```css
+/* makes all classnames available for reference */
+@reference '../../index.module.css';
+
+/* assigns classes to appropriate layer */
+@layer components.l3 {
+
+  .example {
+    /* using apply directive to reference tailwind utility classes */
+    @apply fg-primary-bold px-xs text-body-xs;
+
+    /* using arbitrary css */
+    max-width: 250px;
+
+    /* using pseudo-selectors */
+    &:before {
+      @apply pointer-events-none absolute top-0 left-0;
+      content: '*';
+    }
+
+    /* using built-in tailwind variants */
+    @variant hover {
+      @apply bg-critical-bold outline-critical-bold;
+    }
+
+    /* using custom design toolkit variants */
+    @variant color-serious {
+      @apply bg-serious-muted outline-serious-bold;
+    }
+
+    /* using order to establish precedence with variants */
+    @variant disabled {
+      @apply bg-critical-muted outline-critical-bold;
+    }
+  }
+}
+```
+
+There are a few important things to notice above:
+
+* **@reference** - the use of the `@reference` directive at the top of your css module is required to avoid duplication of applied classes. [Details](https://tailwindcss.com/docs/functions-and-directives#reference-directive)
+
+* **@layer** -- assigns the classes in that group to a layer. In an application-specific component, that would most likely be `l3`.
```

Comment:
to match the other directives, we could include a docs link

```suggestion
* **@layer** -- assigns the classes in that group to a layer. In an application-specific component, that would most likely be `l3`. [Detaiils](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/At-rules/@layer)
```

and maybe we shoudl note that, generally, folks should stick to the already-defined layers, unless they have a pretty good reason to create their own...